scrypto/src/resource/proof.rs

```diff
@@ -96,8 +130,103 @@ impl Proof {
         validated_proof.into()
     }
 
+    fn validate(&self, validation_mode: ProofValidationMode) -> Result<(), ProofValidationError> {
+        match validation_mode {
+            ProofValidationMode::ValidateResourceAddress(resource_address) => {
+                self.validate_resource_address(resource_address)?;
+                Ok(())
+            }
+            ProofValidationMode::ValidateResourceAddressBelongsTo(resource_addresses) => {
+                self.validate_resource_address_belongs_to(&resource_addresses)?;
+                Ok(())
+            }
+            ProofValidationMode::ValidateContainsNonFungible(non_fungible_address) => {
+                self.validate_resource_address(non_fungible_address.resource_address())?;
+                self.validate_contains_non_fungible_id(non_fungible_address.non_fungible_id())?;
+                Ok(())
+            }
+            ProofValidationMode::ValidateContainsNonFungibles(
+                resource_address,
+                non_fungible_ids,
+            ) => {
+                self.validate_resource_address(resource_address)?;
+                self.validate_contains_non_fungible_ids(&non_fungible_ids)?;
+                Ok(())
+            }
+            ProofValidationMode::ValidateContainsAmount(resource_address, amount) => {
+                self.validate_resource_address(resource_address)?;
+                self.validate_contains_amount(amount)?;
+                Ok(())
+            }
+        }
+    }
+
+    fn validate_resource_address(
+        &self,
+        resource_address: ResourceAddress,
+    ) -> Result<(), ProofValidationError> {
+        if self.resource_address() == resource_address {
+            Ok(())
+        } else {
+            Err(ProofValidationError::InvalidResourceAddress(
+                resource_address,
+            ))
+        }
+    }
+
+    fn validate_resource_address_belongs_to(
+        &self,
+        resource_addresses: &BTreeSet<ResourceAddress>,
+    ) -> Result<(), ProofValidationError> {
+        if resource_addresses.contains(&self.resource_address()) {
+            Ok(())
+        } else {
+            Err(ProofValidationError::ResourceAddressDoesNotBelongToList)
+        }
+    }
+
+    fn validate_contains_non_fungible_id(
+        &self,
+        non_fungible_id: NonFungibleId,
+    ) -> Result<(), ProofValidationError> {
+        if self.non_fungible_ids().get(&non_fungible_id).is_some() {
+            Ok(())
+        } else {
+            Err(ProofValidationError::InvalidNonFungibleId(non_fungible_id))
```

Comment:
Minor: suggest renaming to `NonFungibleIdNotFound`

Reply:
@iamyulong Thanks Yulong, will make this change and merge :)